What is the average number of bike rentals per day?

 We still  have to create a dependent variable(number of trips) column in our dataframe. To do that, let's start by creating a tripdate column  in our existing dataframe. We can then group by counting the number of trips per date. Then we can simply do a merge the two dataframes.Let's add a trip date column.

In [35]:
#import necessary libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [36]:
#read the csv file
trip = pd.read_csv("Pittsburgh_healthy_bikeshare.csv", parse_dates = ["start_time","stop_time"])

In [37]:
#add a trip date column
trip['tripdate'] = trip['start_time'].map(lambda x: x.date())

In [38]:
# Create a Group by object using tripdate
grp_date = trip.groupby('tripdate')

# Compute number of trips by date and check out the result
trips_by_date = pd.DataFrame(grp_date.size(), columns=['num_trips_day'])

print(trips_by_date)

            num_trips_day
tripdate                 
2015-05-31            158
2015-06-01             52
2015-06-02             62
2015-06-03             83
2015-06-04            127
2015-06-05            137
2015-06-06            174
2015-06-07            180
2015-06-08             85
2015-06-09            163
2015-06-10            148
2015-06-11            105
2015-06-12            159
2015-06-13            202
2015-06-14            166
2015-06-15            115
2015-06-16            180
2015-06-17            178
2015-06-18            169
2015-06-19            187
2015-06-20            140
2015-06-21            199
2015-06-22            278
2015-06-23            203
2015-06-24            299
2015-06-25            145
2015-06-26            250
2015-06-27            112
2015-06-28            240
2015-06-29            179
...                   ...
2018-12-01             18
2018-12-02             74
2018-12-03             37
2018-12-04             41
2018-12-05             30
2018-12-06  

In [39]:
trips_by_date.head(2)

num_trips_day
tripdate                 
2015-05-31            158
2015-06-01             52

In [40]:
#Since we need to merge the two dataframes, let's reset the index
trips_by_date.reset_index(level=0, inplace=True)
trips_by_date.head(2)

tripdate  num_trips_day
0  2015-05-31            158
1  2015-06-01             52

In [41]:
#Let's merge the two dataframes
merged = pd.merge(trip, trips_by_date, on = "tripdate")

In [42]:
merged.head()

start_time   trip_id           stop_time  bike_id  trip_duration  \
0 2015-05-31 07:18:00  10677999 2015-05-31 07:52:00    70381           2018   
1 2015-05-31 07:25:00  10678335 2015-05-31 07:57:00    70331           1900   
2 2015-05-31 07:53:00  10679549 2015-05-31 09:34:00    70430           6068   
3 2015-05-31 08:05:00  10680090 2015-05-31 08:50:00    70381           2692   
4 2015-05-31 08:06:00  10680133 2015-05-31 08:47:00    70147           2480   

   from_station_id           from_station_name  to_station_id  \
0             1019         42nd St & Butler St           1001   
1             1005       Forbes Ave & Grant St           1000   
2             1050            Healthy Ride Hub           1016   
3             1001  Forbes Ave & Market Square           1017   
4             1001  Forbes Ave & Market Square           1017   

                 station_name   user_type  ...  month  tr_du_mins  tr_du_hr  \
0  Forbes Ave & Market Square  Subscriber  ...      5   33.633333  0.560556   
1    Liberty Ave & Stanwix St    Customer  ...      5   31.666667  0.527778   
2          17th St & Penn Ave  Subscriber  ...      5  101.133333  1.685556   
3          21st St & Penn Ave    Customer  ...      5   44.866667  0.747778   
4          21st St & Penn Ave    Customer  ...      5   41.333333  0.688889   

   speed  row_count  day_of_the_week  start_time_hour  start_time_day_name  \
0   5.41          8           Sunday                7               Sunday   
1   0.80         12           Sunday                7               Sunday   
2   0.75         37           Sunday                7               Sunday   
3   1.73         50           Sunday                8               Sunday   
4   1.87         52           Sunday                8               Sunday   

     tripdate  num_trips_day  
0  2015-05-31            158  
1  2015-05-31            158  
2  2015-05-31            158  
3  2015-05-31            158  
4  2015-05-31            158  

[5 rows x 34 columns]

In [43]:
trip.shape

(163377, 33)

In [44]:
merged.shape

(163377, 34)

In [45]:
merged.num_trips_day.value_counts()

199    2587
163    1956
161    1932
216    1728
215    1720
171    1710
175    1575
170    1530
166    1494
204    1428
176    1408
173    1384
229    1374
149    1341
186    1302
185    1295
214    1284
183    1281
127    1270
211    1266
180    1260
178    1246
133    1197
195    1170
167    1169
285    1140
187    1122
140    1120
181    1086
153    1071
       ... 
20      160
77      154
71      142
68      136
15      135
22      132
16      128
64      128
21      126
31      124
94       94
10       90
88       88
13       78
11       77
19       76
72       72
23       69
69       69
17       68
9        36
14       28
12       24
8        24
6        24
7        21
5        15
3        12
4         8
1         3
Name: num_trips_day, Length: 308, dtype: int64

In [46]:
trips_by_date.num_trips_day.value_counts()

18     16
30     15
25     15
199    13
44     12
161    12
163    12
37     11
24     11
27     11
26     10
36     10
171    10
67     10
127    10
90      9
35      9
149     9
32      9
52      9
58      9
28      9
133     9
10      9
39      9
166     9
170     9
93      9
15      9
175     9
       ..
213     1
94      1
88      1
234     1
239     1
240     1
241     1
242     1
244     1
245     1
249     1
72      1
250     1
252     1
294     1
256     1
258     1
259     1
262     1
264     1
265     1
268     1
272     1
273     1
278     1
283     1
287     1
288     1
290     1
469     1
Name: num_trips_day, Length: 308, dtype: int64

Let's clean up by dropping unnecessary coulumns.

In [47]:
merged.columns

Index(['start_time', 'trip_id', 'stop_time', 'bike_id', 'trip_duration',
       'from_station_id', 'from_station_name', 'to_station_id', 'station_name',
       'user_type', 'TMAX', 'TMIN', 'season', 'holiday', 'walk_score',
       'transit_score', 'bike_score', 'distance', 'event_type', 'rain', 'wind',
       'weather', 'date', 'hour', 'month', 'tr_du_mins', 'tr_du_hr', 'speed',
       'row_count', 'day_of_the_week', 'start_time_hour',
       'start_time_day_name', 'tripdate', 'num_trips_day'],
      dtype='object')

In [48]:
merged.drop(['tr_du_mins', 'tr_du_hr','start_time_day_name','start_time_hour','row_count', 'date'], axis = 1, inplace = True)

In [49]:
merged.columns

Index(['start_time', 'trip_id', 'stop_time', 'bike_id', 'trip_duration',
       'from_station_id', 'from_station_name', 'to_station_id', 'station_name',
       'user_type', 'TMAX', 'TMIN', 'season', 'holiday', 'walk_score',
       'transit_score', 'bike_score', 'distance', 'event_type', 'rain', 'wind',
       'weather', 'hour', 'month', 'speed', 'day_of_the_week', 'tripdate',
       'num_trips_day'],
      dtype='object')

In [50]:
n= merged.to_csv("Healthy_rides_data.csv", index= False)

In [51]:
merged.shape

(163377, 28)

In [52]:
merged.head(2)

start_time   trip_id           stop_time  bike_id  trip_duration  \
0 2015-05-31 07:18:00  10677999 2015-05-31 07:52:00    70381           2018   
1 2015-05-31 07:25:00  10678335 2015-05-31 07:57:00    70331           1900   

   from_station_id      from_station_name  to_station_id  \
0             1019    42nd St & Butler St           1001   
1             1005  Forbes Ave & Grant St           1000   

                 station_name   user_type  ...  event_type  rain  wind  \
0  Forbes Ave & Market Square  Subscriber  ...           N  0.37  8.95   
1    Liberty Ave & Stanwix St    Customer  ...           N  0.37  8.95   

   weather  hour  month  speed  day_of_the_week    tripdate  num_trips_day  
0     good     7      5   5.41           Sunday  2015-05-31            158  
1     good     7      5   0.80           Sunday  2015-05-31            158  

[2 rows x 28 columns]

In [53]:
trip["from_station_id"] = trip["from_station_id"].astype(str)
trip.dtypes

start_time             datetime64[ns]
trip_id                         int64
stop_time              datetime64[ns]
bike_id                         int64
trip_duration                   int64
from_station_id                object
from_station_name              object
to_station_id                   int64
station_name                   object
user_type                      object
TMAX                          float64
TMIN                          float64
season                         object
holiday                         int64
walk_score                    float64
transit_score                 float64
bike_score                    float64
distance                      float64
event_type                     object
rain                          float64
wind                          float64
weather                        object
date                           object
hour                            int64
month                           int64
tr_du_mins                    float64
tr_du_hr    

In [54]:
def daily_avg_trips(station_id):
    # Group by tripdate in the trip dataframe
    grp_date = trip[(trip.from_station_id == station_id)].groupby(['tripdate'])

    # Compute number of trips by date and check out the result
    trips_by_date = pd.DataFrame(grp_date.size(), columns=['num_trips'])
    trips_by_date['weekday'] = trips_by_date.index.map(lambda x: x.weekday())

    # Compute the average
    mean_trips = trips_by_date['num_trips'].mean()
    tot_tripdates = trips_by_date['num_trips'].count()
    num_days = (trips_by_date.index.max() - trips_by_date.index.min()).days + 1

    print("The mean number of trips per day is {:.2f} out of {} tripdates."
          .format(mean_trips, tot_tripdates))
    print(f"Date range is from {trips_by_date.index.min()} : {trips_by_date.index.max()} ")
    print("There are {} days in the date range.".format(num_days))
    
    return trips_by_date


In [55]:
u = daily_avg_trips('1019')

The mean number of trips per day is 4.27 out of 953 tripdates.
Date range is from 2015-05-31 : 2018-12-27 
There are 1307 days in the date range.


Notice that the mean uses a denominator of 953 even though there are 1307 days in the date range of interest. Zero days were not accounted for.
We will rectify iy by craeting an empty dataframe based range of dates from the start till end date of the time period.

In [56]:
# Create range of dates to use as an index
start = pd.datetime(2015, 5, 31)
end = pd.datetime(2018, 12, 27)
rng = pd.date_range(start, end) 
print(rng)

#Create an empty DataFrame using the range of dates as the index.
trips_by_date_new = pd.DataFrame(index=rng)

## Add weekday column to new dataframe
trips_by_date_new['weekday'] = trips_by_date_new.index.map(lambda x: x.weekday())
print("\n")
print("trips_by_date_new dataframe:")
trips_by_date_new.head()

DatetimeIndex(['2015-05-31', '2015-06-01', '2015-06-02', '2015-06-03',
               '2015-06-04', '2015-06-05', '2015-06-06', '2015-06-07',
               '2015-06-08', '2015-06-09',
               ...
               '2018-12-18', '2018-12-19', '2018-12-20', '2018-12-21',
               '2018-12-22', '2018-12-23', '2018-12-24', '2018-12-25',
               '2018-12-26', '2018-12-27'],
              dtype='datetime64[ns]', length=1307, freq='D')


trips_by_date_new dataframe:


weekday
2015-05-31        6
2015-06-01        0
2015-06-02        1
2015-06-03        2
2015-06-04        3

In [57]:
#Use groupby on the original trip DataFrame to compute number of trips by date
# Create a Group by object using tripdate
grp_date = trip.groupby(['tripdate'])

# Compute number of trips by date and check out the result
trips_by_date = pd.DataFrame(grp_date.size(), columns=['num_trips'])

# Merge the two DataFrames on their indexes (tripdate) but do a "left join" with new dataframe on left.
trips_by_date_merged = pd.merge(trips_by_date_new, trips_by_date, how='left', left_index=True, right_index=True, sort=True)


#Replace missing values with zeroes for those dates with no trips
# Fill in any missing values with 0.
trips_by_date_merged['num_trips'] = trips_by_date_merged['num_trips'].fillna(0)

print(trips_by_date_merged.head())

            weekday  num_trips
2015-05-31        6      158.0
2015-06-01        0       52.0
2015-06-02        1       62.0
2015-06-03        2       83.0
2015-06-04        3      127.0


In [58]:
# Compute the average
mean_trips = trips_by_date_merged['num_trips'].mean()
tot_tripdates = trips_by_date_merged['num_trips'].count() 
num_days = (trips_by_date_merged.index.max() - trips_by_date_merged.index.min()).days + 1

print("The mean number of trips per day is {:.2f} out of  {} tripdates.".format(mean_trips, tot_tripdates))
print(f"Date range is from {trips_by_date.index.min()} : {trips_by_date.index.max()} ")
print("There are {} days in the date range.".format(num_days))

The mean number of trips per day is 124.95 out of  1307 tripdates.
Date range is from 2015-05-31 : 2018-12-30 
There are 1307 days in the date range.


In [59]:
#Average number of rentals per day at individual station including zero days:

def daily_avg_trips_new(station_id, start_date = pd.datetime(2015, 5, 31), end_date = pd.datetime(2018, 12, 27)):
    
    # Step 1: Create range of dates based on analysis timeframe
    rng = pd.date_range(start_date, end_date)
    
    # Step 2: Create an empty DataFrame using the range of dates as the index.
    trips_by_date_new = pd.DataFrame(index=rng)
    # Add weekday column to new dataframe
    trips_by_date_new['weekday'] = trips_by_date_new.index.map(lambda x: x.weekday())
    
    # Step 3: Use groupby on the original trip DataFrame to compute number of trips by date
    grp_date = trip[(trip.from_station_id == station_id)].groupby(['tripdate'])
    # Compute number of trips by date 
    trips_by_date = pd.DataFrame(grp_date.size(), columns=['num_trips'])

    # Step 4: Merge the two DataFrames on their indexes (tripdate) but do a "left join". 
    trips_by_date_merged = pd.merge(trips_by_date_new, trips_by_date, how='left', left_index=True, right_index=True, sort=True)
    
    # Step 5: Replace missing values with zeroes for those dates with no trips
    trips_by_date_merged['num_trips'] = trips_by_date_merged['num_trips'].fillna(0)
    
    # Compute the average
    mean_trips = trips_by_date_merged['num_trips'].mean()
    tot_tripdates = trips_by_date_merged['num_trips'].count() 
    num_days = (trips_by_date_merged.index.max() - trips_by_date_merged.index.min()).days + 1

    print("The mean number of trips per day is {:.2f} out of  {} tripdates.".format(mean_trips, tot_tripdates))
    print(f"Date range is from {trips_by_date.index.min()} : {trips_by_date.index.max()} ")
    print("There are {} days in the date range.".format(num_days))
    
    return trips_by_date_merged
    
    
    

In [60]:
i = daily_avg_trips_new('1019')

The mean number of trips per day is 3.11 out of  1307 tripdates.
Date range is from 2015-05-31 : 2018-12-27 
There are 1307 days in the date range.


In [61]:
u = daily_avg_trips('1019')

The mean number of trips per day is 4.27 out of 953 tripdates.
Date range is from 2015-05-31 : 2018-12-27 
There are 1307 days in the date range.
